### PR TITLE
fix: thread leaking

### LINKED ISSUE #10374: bug: Query crashed during datax importing process
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

0.9.58

### What's Wrong?

error log shows `Out of memory (os error 12)`, but the memory usage is not high.

It happend twice:
1.
image
2.
image

The error log prior to the crash
```
<Backtrace disabled by default. Please use RUST_BACKTRACE=1 to enable> 
2023-03-06T04:18:11.113672Z ERROR databend_query::servers::http::middleware: src/query/service/src/servers/http/middleware.rs:196: http request error: query id not found 90be055c-8538-4f3a-aee6-aeffd7a627a5
2023-03-06T04:19:23.953077Z ERROR databend_query::servers::http::middleware: src/query/service/src/servers/http/middleware.rs:196: http request error: query id not found 851a5f7c-e1e0-4caf-b44d-3fa5d8d00d16
2023-03-06T04:21:21.417970Z ERROR common_tracing::panic_hook: src/common/tracing/src/panic_hook.rs:38: panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', /workspace/src/common/base/src/runtime/thread.rs:85:18 backtrace=Backtrace [{ fn: "common_tracing::panic_hook::log_panic", file: "/workspace/src/common/tracing/src/panic_hook.rs", line: 35 }, { fn: "common_tracing::panic_hook::set_panic_hook::{{closure}}", file: "/workspace/src/common/tracing/src/panic_hook.rs", line: 30 }, { fn: "<alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/alloc/src/boxed.rs", line: 2032 }, { fn: "std::panicking::rust_panic_with_hook", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 692 }, { fn: "std::panicking::begin_panic_handler::{{closure}}", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 579 }, { fn: "std::sys_common::backtrace::__rust_end_short_backtrace", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys_common/backtrace.rs", line: 137 }, { fn: "rust_begin_unwind", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 575 }, { fn: "core::panicking::panic_fmt", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/panicking.rs", line: 64 }, { fn: "core::result::unwrap_failed", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/result.rs", line: 1790 }, { fn: "core::result::Result<T,E>::unwrap", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/result.rs", line: 1112 }, { fn: "common_base::runtime::thread::Thread::named_spawn", file: "/workspace/src/common/base/src/runtime/thread.rs", line: 77 }, { fn: "databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads", file: "/workspace/src/query/service/src/pipelines/executor/pipeline_executor.rs", line: 293 }, { fn: "databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute", file: "/workspace/src/query/service/src/pipelines/executor/pipeline_executor.rs", line: 186 }, { fn: "databend_query::pipelines::executor::pipeline_pulling_executor::PipelinePullingExecutor::thread_function::{{closure}}", file: "/workspace/src/query/service/src/pipelines/executor/pipeline_pulling_executor.rs", line: 177 }, { fn: "core::ops::function::FnOnce::call_once", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/ops/function.rs", line: 507 }, { fn: "common_base::runtime::thread::Thread::named_spawn::{{closure}}", file: "/workspace/src/common/base/src/runtime/thread.rs", line: 83 }, { fn: "std::sys_common::backtrace::__rust_begin_short_backtrace", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys_common/backtrace.rs", line: 121 }, { fn: "std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/thread/mod.rs", line: 550 }, { fn: "<core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/panic/unwind_safe.rs", line: 271 }, { fn: "std::panicking::try::do_call", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 483 }, { fn: "std::panicking::try", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 447 }, { fn: "std::panic::catch_unwind", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panic.rs", line: 137 }, { fn: "std::thread::Builder::spawn_unchecked_::{{closure}}", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/thread/mod.rs", line: 549 }, { fn: "core::ops::function::FnOnce::call_once{{vtable.shim}}", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/ops/function.rs", line: 507 }, { fn: "<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/alloc/src/boxed.rs", line: 2000 }, { fn: "<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/alloc/src/boxed.rs", line: 2000 }, { fn: "std::sys::unix::thread::Thread::new::thread_start", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/thread.rs", line: 108 }] panic.file="/workspace/src/common/base/src/runtime/thread.rs" panic.line=85 panic.column=18
2023-03-06T04:21:21.420977Z ERROR common_tracing::panic_hook: src/common/tracing/src/panic_hook.rs:38: panicked at 'failed to set up alternative stack guard page: Out of memory (os error 12)', library/std/src/sys/unix/stack_overflow.rs:147:13 backtrace=Backtrace [{ fn: "common_tracing::panic_hook::log_panic", file: "/workspace/src/common/tracing/src/panic_hook.rs", line: 35 }, { fn: "common_tracing::panic_hook::set_panic_hook::{{closure}}", file: "/workspace/src/common/tracing/src/panic_hook.rs", line: 30 }, { fn: "<alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/alloc/src/boxed.rs", line: 2032 }, { fn: "std::panicking::rust_panic_with_hook", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 692 }, { fn: "std::panicking::begin_panic_handler::{{closure}}", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 579 }, { fn: "std::sys_common::backtrace::__rust_end_short_backtrace", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys_common/backtrace.rs", line: 137 }, { fn: "rust_begin_unwind", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/panicking.rs", line: 575 }, { fn: "core::panicking::panic_fmt", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/core/src/panicking.rs", line: 64 }, { fn: "std::sys::unix::stack_overflow::imp::get_stackp", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/stack_overflow.rs", line: 147 }, { fn: "std::sys::unix::stack_overflow::imp::get_stack", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/stack_overflow.rs", line: 153 }, { fn: "std::sys::unix::stack_overflow::imp::make_handler", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/stack_overflow.rs", line: 164 }, { fn: "std::sys::unix::stack_overflow::Handler::new", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/stack_overflow.rs", line: 14 }, { fn: "std::sys::unix::thread::Thread::new::thread_start", file: "/rustc/b70baa4f922a1809d79caeaeb902800c3be283b9/library/std/src/sys/unix/thread.rs", line: 106 }] panic.file="library/std/src/sys/unix/stack_overflow.rs" panic.line=147 panic.column=13
```

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

to avoid deadlock during destructing the ad-hoc runtime `query-ctx`
- introduce `GlobalQueryRuntime`
- replace the ad-hoc runtime `query-ctx` with global runtime `GlobalQueyRuntime` in `HttpQuery::try_create` 
- replace the ad-hoc runtime `query-ctx` with global runtime `GlobalQueyRuntime` in `StatisticsSender::start` 

p.s. This is more like a temp workaround, we should refactor the execution model (the thread pools, ad-hoc runtimes, and communications between sync/async components).

  

Closes #10374

### CODE REVIEW

File: src/common/base/src/runtime/global_runtime.rs [hunk 1]
```diff
@@ -21,6 +21,15 @@ use crate::runtime::Runtime;
 
 pub struct GlobalIORuntime;
 
+pub struct GlobalQueryRuntime(pub Runtime);
+
+impl GlobalQueryRuntime {
+    #[inline(always)]
+    pub fn runtime<'a>(self: &'a Arc<Self>) -> &'a Runtime {
```
Comment (zhang2014) [inline src/common/base/src/runtime/global_runtime.rs:28]: Maybe `impl Deref` is better

Comment (dantengsky) [reply src/common/base/src/runtime/global_runtime.rs:28]: `Deref should only be implemented for smart pointers `

https://doc.rust-lang.org/std/ops/trait.Deref.html

**Comment by BohuTANG** [comment]
This test is weird:
```
00_0002_copy_from_fs_location:                                          [ FAIL ] - result differs with:
--- /runner/_work/databend/databend/tests/suites/1_stateful/00_copy/00_0002_copy_from_fs_location.result	2023-03-15 00:45:00.470028456 +0000
+++ /runner/_work/databend/databend/tests/suites/1_stateful/00_copy/00_0002_copy_from_fs_location.stdout	2023-03-15 00:48:18.020775879 +0000
@@ -5,4 +5,4 @@
 199	2020.0	769
 199	2020.0	769
 199	2020.0	769
-199	2020.0	769
+198	2020.0	765
```

Also cc @youngsofun

**Comment by youngsofun** [comment]
>



> This test is weird:
> 
> ```
> 00_0002_copy_from_fs_location:                                          [ FAIL ] - result differs with:
> --- /runner/_work/databend/databend/tests/suites/1_stateful/00_copy/00_0002_copy_from_fs_location.result	2023-03-15 00:45:00.470028456 +0000
> +++ /runner/_work/databend/databend/tests/suites/1_stateful/00_copy/00_0002_copy_from_fs_location.stdout	2023-03-15 00:48:18.020775879 +0000
> @@ -5,4 +5,4 @@
>  199	2020.0	769
>  199	2020.0	769
>  199	2020.0	769
> -199	2020.0	769
> +198	2020.0	765
> ```
> 
> Also cc @youngsofun

I remember meet this before this pr,  not related if rerun success I think,
I will check it further.

**Review by BohuTANG — APPROVED** [review]

File: src/common/base/src/runtime/global_runtime.rs (hunk 1 above)
Comment (BohuTANG) [reply src/common/base/src/runtime/global_runtime.rs:28]: I will mark this discussion as solved and prepare to merge this PR.

**Pull request merged**